### PR TITLE
Mechs rememeber removed cell's charge level and adjust inserted one's accordingly

### PR DESCRIPTION
**Pull request by X-TheDark**
As suggested to be added by @Bawhoppen 

This allows you to upgrade the mech's cell without it being drained.

When a power cell is removed from the mech, it remembers the cell's remaining power percentage and when a new one is inserted, it adjusts it's charge level to that percentage, if it is higher (if inserted cells charge percent < removed cell's -> no change).

It never decreases the power cells level below 10% (in case you remove empty cell).

Also removes the (what I believe to be) unnecessary `src` everywhere.

:cl: X-TheDark
rscadd: Mech battery replacement improved. The inserted battery's charge will be set to the same percentage of the removed one's (or 10%, if removed one's was less).
/:cl:

### CODE REVIEW

File: code/game/mecha/mecha_defense.dm [hunk 1]
```diff
@@ -227,48 +227,49 @@
 		if(internal_damage & MECHA_INT_TEMP_CONTROL)
 			clearInternalDamage(MECHA_INT_TEMP_CONTROL)
 			user << "<span class='notice'>You repair the damaged temperature controller.</span>"
-		else if(state==3 && src.cell)
-			src.cell.forceMove(src.loc)
-			src.cell = null
+		else if(state==3 && cell)
+			cell_power_remaining = max(0.1, cell.charge/cell.maxcharge) //10% charge or whatever is remaining in the current cell
```
Comment (X-TheDark) [inline code/game/mecha/mecha_defense.dm]: Thus is the relevant line of code among all the `src` removal.

File: code/game/mecha/mecha_defense.dm [hunk 2]
```diff
@@ -227,48 +227,49 @@
...
+			if(!cell)
 				if(!user.drop_item())
 					return
 				var/obj/item/weapon/stock_parts/cell/C = W
 				user << "<span class='notice'>You install the powercell.</span>"
 				C.forceMove(src)
-				C.use(C.charge * 0.8)
-				src.cell = C
-				src.log_message("Powercell installed")
+				C.use(max(0, C.charge - C.maxcharge*cell_power_remaining)) //Set inserted cell's power to saved percentage if that's higher
```
Comment (X-TheDark) [inline code/game/mecha/mecha_defense.dm]: And this one.

**Comment by Bawhoppen** [comment]
fyi to everyone this is another yog one so its exempt from permafrost policy (plus its a very small refactor I guess)

**Comment by X-TheDark** [comment]
I don't mind doing the whole Merge Token thing, but don't expect me to do it any time soon, I'm afraid. Bit too busy with the rebase.

**Comment by Jazaen** [comment]
Does this mean that ifa mech gets stranded somewhere because it ran out of power you czn't recharge it? (Though I am not sure you could do that in the first place)

**Comment by X-TheDark** [comment]
No, it always leaves at least 10% charge in the powercell. Basically, if you insert a new powercell to replace the completely empty one, it'll keep 10% charge (if you wish, I could change it back to 20%, since originally, it used 80% of the charge **every time you inserted the cell**).

This is mostly a QoL thing for upgrading cells.

**Comment by GunHog** [comment]
Personally, I would have just removed the cell drain mechanic altogether, but this is better than it currently is...

**Comment by xxalpha** [comment]
Why drain at all?

**Comment by KorPhaeron** [comment]
Probably to avoid mech pilots just swapping cells out in the field

**Comment by GunHog** [comment]
Field swapping cells leaves quite a window of vulnerability, I believe @Razharas added this mechanic as a way of forcing the pilot to retreat to a mech recharger.

Please note the following steps required for a cell swap:
- Check that your mech is set to allow maintenance mode.
- Eject from the mech.
- Swipe your ID on the mech to open its interface.
- Activate maintenance mode on the mech.
- Wrench
- Crowbar
- Apply screwdriver to the mech in order to remove the cell.
- Insert a new cell.
- Screwdriver
- Crowbar
- Wrench
- Cancel maintenance mode.
- Re-enter the mech.
- (Optional) Disable maintenance mode access

**Comment by X-TheDark** [comment]
Yeah, I'm pretty sure it's a balance thing. Replacing the 0% cell with 100% is still gorillion times faster than finding one of the fixed amount of recharge pads that exist on the station and standing there and waiting for it to recharge to that level by itself.

**Comment by Supermichael777** [comment]
raz hates fun reverting his shirt balance memes should be free

**Comment by optimumtact** [comment]
Yog, more like YOGHURT

**Pull request merged**